test/ringbuffer-ut.cc

```diff
@@ -251,3 +251,134 @@ TEST(ringbuffer, full) {
   EXPECT_TRUE(
       std::equal(buffer.begin(), buffer.end(), buffer2.begin(), buffer2.end()));
 }
+
+void mpsc_reader_fun(RingBuffer *rb, size_t nb_elements, size_t nb_producers,
+                     bool use_new_object, bool advance_eagerly) {
+  std::optional<MPSCRingBufferReader> reader(*rb);
+
+  std::vector<size_t> counts(nb_producers);
+  size_t total_count = 0;
+  while (total_count < nb_elements * nb_producers) {
+    if (use_new_object) {
+      reader.emplace(*rb);
```

Comment:
Minor: I had to admit I had to look up [emplace](https://www.macmillandictionary.com/dictionary/british/emplace) in the dictionary 😅

Reply:
lol, also https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2008/n2642.pdf

But fundamentally, the C++ standards committee has a precise technical goal for using the term "emplace" to qualify a particular type of insertion.

Reply:
Someone wrote 8 pages on this matter....

/me backs away slowly and silently

Reply:
C++ is too complicated for its own good !
The idea behind `emplace` is to be able to construct an object in-place in a container, potentially saving the construction / copy of a temporary object.